### PR TITLE
FEATURE: add `$skipErrors` option to `GlobIterator`

### PR DESCRIPTION
**Pull request by zobo**

### CODE REVIEW

File: src/Iterator/GlobIterator.php [hunk 1]
```diff
@@ -33,9 +33,12 @@ class GlobIterator extends IteratorIterator
...
      * @param string $glob  The glob pattern.
      * @param int    $flags A bitwise combination of the flag constants in
-     *                      {@link Glob}.
+     *               {@link Glob}.
+     * @param bool   $skip_errors Add the RecursiveIteratorIterator::CATCH_GET_CHILD
+     *               to internal RecursiveIteratorIterator and prevent throwing
+     *               Exception on errors like access denied.
      */
-    public function __construct($glob, $flags = 0)
+    public function __construct($glob, $flags = 0, $skip_errors = false)
```
Comment (Ocramius) [inline src/Iterator/GlobIterator.php]: This library requires PHP 7.3 or newer, so we can safely use `bool` here

Comment (Ocramius) [reply src/Iterator/GlobIterator.php]: Let's also call this `$skipErrors`

**Review by Ocramius — CHANGES_REQUESTED** [review]
Requires an automated test to validate the behavior of the new flag and prevent regressions long-term.

File: src/Iterator/GlobIterator.php (hunk 1 above)
Comment (zobo) [reply src/Iterator/GlobIterator.php]: Agree will push an update shortly.

**Comment by zobo** [comment]
I can imagine hacking something for linux based tests. chmod-ing a directory and so causing an access error...

**Comment by Ocramius** [comment]
Yup, pretty much what is needed 👍

File: src/Iterator/GlobIterator.php (hunk 1 above)
Comment (zobo) [reply src/Iterator/GlobIterator.php]: I also added types for the other two arguments.

**Comment by zobo** [comment]
Added a test also.

Amazingly, I managed to screw up the RecursiveIteratorIterator flag (because I copied the code around so much) and the test caught it :)

File: tests/Iterator/GlobIteratorErrorTest.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    public function testIterateError()
+    {
+        if (defined('PHP_WINDOWS_VERSION_MAJOR')) {
+            $this->markTestSkipped('chmod tests only on Linux.');
+
+            return;
+        }
+        $this->expectException(\UnexpectedValueException::class);
```
Comment (Ocramius) [inline tests/Iterator/GlobIteratorErrorTest.php:55]: Let's move this expectation right before the statement that is supposed to throw

**Review by Ocramius — CHANGES_REQUESTED** [review]
Tests currently failing 😁

File: tests/Iterator/GlobIteratorErrorTest.php (hunk 2 above)
Comment (zobo) [reply tests/Iterator/GlobIteratorErrorTest.php:55]: I made a mistake in the arguments... I combined flags with mode... I must really need some sleep :D

Comment (zobo) [reply tests/Iterator/GlobIteratorErrorTest.php:55]: Hm, all tests pass here, but I noticed the mistake when I included the code in vscode-php-intellisense ...
Can you make the workflow run on this repo?

Comment (zobo) [reply tests/Iterator/GlobIteratorErrorTest.php:55]: Ah thanks. Will look at the tests again shortly... I'll try to run them in a VM instead of WSL.

Comment (zobo) [reply tests/Iterator/GlobIteratorErrorTest.php:55]: I think its failing because some checks run under root (alpine docker?). Added logic to skip test in that case. Please re-run.

Comment (zobo) [reply tests/Iterator/GlobIteratorErrorTest.php:55]: All tests are OK now. I checked that they actually get executed where it matters. Linux and macos - where the current user isn't root.

**Comment by zobo** [comment]
Now the only other change I can think of is exposing $skipErrors to `Glob::glob`.

File: src/Iterator/GlobIterator.php [hunk 3]
```diff
@@ -33,9 +33,12 @@ class GlobIterator extends IteratorIterator
...
      * @param string $glob  The glob pattern.
      * @param int    $flags A bitwise combination of the flag constants in
-     *                      {@link Glob}.
+     *               {@link Glob}.
+     * @param bool   $skipErrors Add the RecursiveIteratorIterator::CATCH_GET_CHILD
+     *               to internal RecursiveIteratorIterator and prevent throwing
+     *               Exception on errors like access denied.
      */
-    public function __construct($glob, $flags = 0)
+    public function __construct(string $glob, int $flags = 0, bool $skipErrors = false)
```
Comment (Ocramius) [inline src/Iterator/GlobIterator.php]: Hmm, we can't really enforce `string $glob` and `int $flags` here, as it would break BC for downstream consumers :thinking: 

The type additions can only really be safely added to the new constructor argument that you added

File: tests/Iterator/GlobIteratorErrorTest.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    public function testIterateWithoutError()
+    {
+        if (defined('PHP_WINDOWS_VERSION_MAJOR')) {
+            $this->markTestSkipped('chmod tests only on Linux.');
+
+            return;
+        }
+        $this->assertTrue(function_exists('posix_getuid'));
+        if (posix_getuid() === 0) {
```
Comment (Ocramius) [inline tests/Iterator/GlobIteratorErrorTest.php:69]: Is `0` always `root` in posix-compliant systems?

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: tests/Iterator/GlobIteratorErrorTest.php (hunk 4 above)
Comment (zobo) [reply tests/Iterator/GlobIteratorErrorTest.php:69]: Yes. You could have normal users with higher privileges, but for this edge case it should suffice.

File: src/Iterator/GlobIterator.php (hunk 3 above)
Comment (zobo) [reply src/Iterator/GlobIterator.php]: I agree. This could break BC.

**Review by Ocramius — APPROVED** [review]
Thanks @zobo!
